### DISCUSSION TITLE
pyOpenMS dealing with common errors

### DISCUSSION THREAD

**Discussion by timosachsenberg** [OpenMS/OpenMS#5275] [Q&A]
- What happens if two functions with same name (e.g., toString are generated example: DataValue)
- What if two enums have the same name but are internal classes / other namespace? e.g., ParamValue::DataType and DataValue::DataType
- Output error message if pxd comment is found that can't be parsed (e.g., extra space after : )
- Python exceptions during code generation seem to be swallowed ? is there a rethrow needed?
- Maybe not necessary but could e.g., pandas be included?
- String handling:
  - string references are not correctly handled
  - unify and rework string encoding/conversion between python and our C++ classes

**Comment by timosachsenberg** [comment]
We probably need a way to handle redeclaration.
warning: /home/sachsenb/OMS/OpenMS/src/pyOpenMS/pxds/DataValue.pxd:47:8: 'STRING_VALUE' redeclared
warning: /home/sachsenb/OMS/OpenMS/src/pyOpenMS/pxds/DataValue.pxd:48:8: 'INT_VALUE' redeclared
warning: /home/sachsenb/OMS/OpenMS/src/pyOpenMS/pxds/DataValue.pxd:49:8: 'DOUBLE_VALUE' redeclared
warning: /home/sachsenb/OMS/OpenMS/src/pyOpenMS/pxds/DataValue.pxd:50:8: 'STRING_LIST' redeclared
warning: /home/sachsenb/OMS/OpenMS/src/pyOpenMS/pxds/DataValue.pxd:51:8: 'INT_LIST' redeclared
warning: /home/sachsenb/OMS/OpenMS/src/pyOpenMS/pxds/DataValue.pxd:52:8: 'DOUBLE_LIST' redeclared
warning: /home/sachsenb/OMS/OpenMS/src/pyOpenMS/pxds/DataValue.pxd:53:8: 'EMPTY_VALUE' redeclared
warning: /home/sachsenb/OMS/OpenMS/src/pyOpenMS/pxds/ControlledVocabulary.pxd:80:8: 'NONE' redeclared
warning: /home/sachsenb/OMS/OpenMS/src/pyOpenMS/pxds/SpectrumSettings.pxd:54:8: 'UNKNOWN' redeclared
warning: /home/sachsenb/OMS/OpenMS/src/pyOpenMS/pxds/MassAnalyzer.pxd:167:6: 'NONE' redeclared
warning: /home/sachsenb/OMS/OpenMS/src/pyOpenMS/pxds/TargetedExperimentHelper.pxd:16:12: 'UNKNOWN' redeclared
warning: /home/sachsenb/OMS/OpenMS/src/pyOpenMS/pxds/TargetedExperimentHelper.pxd:29:12: 'UNKNOWN' redeclared
warning: /home/sachsenb/OMS/OpenMS/src/pyOpenMS/pxds/ReactionMonitoringTransition.pxd:74:8: 'UNKNOWN' redeclared


potential solutions? https://stackoverflow.com/questions/31001918/wrap-enum-class-for-cython

**Reply by timosachsenberg** [reply]
https://github.com/OpenMS/OpenMS/issues/4505

**Reply by jpfeuffer** [reply]
maybe here: https://github.com/uweschmitt/autowrap/blob/2dcf9724294f23f87ae686f7d5a5f193a5fab3e2/autowrap/CodeGenerator.py#L308

**Comment by timosachsenberg** [comment]
What do these mean? outdated?
At global scope:
cc1plus: warning: unrecognized command line option ‘-Wno-deprecated-copy’
cc1plus: warning: unrecognized command line option ‘-Wno-header-guard’
cc1plus: warning: unrecognized command line option ‘-Wno-deprecated-register’
cc1plus: warning: unrecognized command line option ‘-Wno-redeclared-class-member’

**Reply by jpfeuffer** [reply]
yes outdated or platform specific

**Comment by timosachsenberg** [comment]
do these need to be warnigns?

removing build/bdist.linux-x86_64/wheel
running bdist
running bdist_dumb
running build
running build_py
copying pyopenms/version.py -> build/lib.linux-x86_64-3.8/pyopenms
running egg_info
writing pyopenms.egg-info/PKG-INFO
writing dependency_links to pyopenms.egg-info/dependency_links.txt
writing requirements to pyopenms.egg-info/requires.txt
writing top-level names to pyopenms.egg-info/top_level.txt
reading manifest file 'pyopenms.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching 'pyopenms/*.dll'
warning: no files found matching 'pyopenms/*.dylib'
warning: no directories found matching 'pyopenms/*.framework'
warning: no files found matching 'pyopenms/Qt*'
warning: no previously-included files matching 'pyopenms.cpp' found anywhere in distribution

**Reply by jpfeuffer** [reply]
can be fixed in the MANIFEST for setuptools but they will be platform-specific as well.

**Comment by timosachsenberg** [comment]
String handling is still confusing. Why is b" needed, "string".encode()? Can we do something about that or make it at least consistent.
How does this play with char*, string, String function/constructor parameter etc.?

**Reply by timosachsenberg** [reply]
see https://github.com/OpenMS/OpenMS/discussions/5218

**Reply by jpfeuffer** [reply]
https://github.com/uweschmitt/autowrap/blob/master/autowrap/ConversionProvider.py#L1501

I think bytes is used as the unique equivalent for string. Not sure how you would handle multiple equivalents though.
Put an empty string there? Is it accepted as a PyObject then, where it needs checking code with "if isinstance()"?

**Comment by timosachsenberg** [comment]
Can we do better than this error message?
```
Exception                                 Traceback (most recent call last)
<ipython-input-4-31bbf18ef2ec> in <module>
----> 1 par.setValue("test", 0.05, "test")
pyopenms/pyopenms_3.pyx in pyopenms.pyopenms_3.Param.setValue()
Exception: can not handle type of ('test', 0.05, 'test')
```
e.g. list signature of existing ones?

**Reply by jpfeuffer** [reply]
https://github.com/uweschmitt/autowrap/blob/2dcf9724294f23f87ae686f7d5a5f193a5fab3e2/autowrap/CodeGenerator.py#L588

You can list the results of self.cr.get(t) somehow.

**Comment by timosachsenberg** [comment]
Document how to wrap classes with e.g. deleted default constructor

**Comment by timosachsenberg** [comment]
Autowrap: what happens if two conversion providers for the same base type are provided? Can we fail with an error?
Document place how converters are added:
src/OpenMS/converters/__init__.py
and
src/OpenMS/converters/special_autowrap_conversionproviders.py

**Reply by jpfeuffer** [reply]
https://github.com/uweschmitt/autowrap/blob/master/autowrap/ConversionProvider.py#L119

Always the last is taken, to allow overwriting with custom providers.

**Reply by timosachsenberg** [reply]
thanks... so probably the right way would be to test this in our init . py (or distinguis build in from custon in autowrap)

**Comment by timosachsenberg** [comment]
Include pyopenms-extra documentation as a submodule, put code into external files and include in doc and then test it as part of the OpenMS pyopenms tests.

**Comment by timosachsenberg** [comment]
E.g. wrapping static functions as non static ones doesn't cause an error. Wrapped objects just don't work
https://github.com/OpenMS/OpenMS/issues/5292

**Comment by timosachsenberg** [comment]
DONE: Exceptions in C++ are now derived from std::runtime_exception and thus forwarded to python see https://github.com/OpenMS/OpenMS/pull/5220
